vpr/src/route/router_lookahead_cost_map.cpp

```diff
@@ -0,0 +1,408 @@
+static constexpr float PENALTY_FACTOR = 1.f;
+static constexpr float PENALTY_MIN = 1e-12f;
+
+// also known as the L1 norm
+static int manhattan_distance(const vtr::Point<int>& a, const vtr::Point<int>& b) {
+    return abs(b.x() - a.x()) + abs(b.y() - a.y());
+}
+
+template<class T>
+static constexpr const T& clamp(const T& v, const T& lo, const T& hi) {
+    return std::min(std::max(v, lo), hi);
+}
+
+template<typename T>
+static vtr::Point<T> closest_point_in_rect(const vtr::Rect<T>& r, const vtr::Point<T>& p) {
+    if (r.empty()) {
+        return vtr::Point<T>(0, 0);
+    } else {
+        return vtr::Point<T>(clamp<T>(p.x(), r.xmin(), r.xmax() - 1),
+                             clamp<T>(p.y(), r.ymin(), r.ymax() - 1));
+    }
+}
+
+// build the segment map
+void CostMap::build_segment_map() {
+    const auto& device_ctx = g_vpr_ctx.device();
+    segment_map_.resize(device_ctx.rr_nodes.size());
+    for (size_t i = 0; i < segment_map_.size(); ++i) {
+        auto& from_node = device_ctx.rr_nodes[i];
+
+        int from_cost_index = from_node.cost_index();
+        int from_seg_index = device_ctx.rr_indexed_data[from_cost_index].seg_index;
+
+        segment_map_[i] = from_seg_index;
+    }
+}
+
+// resize internal data structures
+void CostMap::set_counts(size_t seg_count) {
+    cost_map_.clear();
+    offset_.clear();
+    penalty_.clear();
+    cost_map_.resize({1, seg_count});
+    offset_.resize({1, seg_count});
+    penalty_.resize({1, seg_count});
+    seg_count_ = seg_count;
+}
+
+// cached node -> segment map
+int CostMap::node_to_segment(int from_node_ind) const {
+    return segment_map_[from_node_ind];
+}
+
+static util::Cost_Entry penalize(const util::Cost_Entry& entry, int distance, float penalty) {
+    penalty = std::max(penalty, PENALTY_MIN);
+    return util::Cost_Entry(entry.delay + distance * penalty * PENALTY_FACTOR,
+                            entry.congestion, entry.fill);
+}
+
+// get a cost entry for a segment type, chan type, and offset
+util::Cost_Entry CostMap::find_cost(int from_seg_index, int delta_x, int delta_y) const {
+    VTR_ASSERT(from_seg_index >= 0 && from_seg_index < (ssize_t)offset_.size());
+    const auto& cost_map = cost_map_[0][from_seg_index];
+    if (cost_map.dim_size(0) == 0 || cost_map.dim_size(1) == 0) {
+        return util::Cost_Entry();
+    }
+
+    vtr::Point<int> coord(delta_x - offset_[0][from_seg_index].first,
+                          delta_y - offset_[0][from_seg_index].second);
+    vtr::Rect<int> bounds(0, 0, cost_map.dim_size(0), cost_map.dim_size(1));
+    auto closest = closest_point_in_rect(bounds, coord);
+    auto cost = cost_map_[0][from_seg_index][closest.x()][closest.y()];
+    float penalty = penalty_[0][from_seg_index];
+    auto distance = manhattan_distance(closest, coord);
+    return penalize(cost, distance, penalty);
+}
+
+// set the cost map for a segment type and chan type, filling holes
```

Comment:
Adding the in-code comment. For the time being, the algorithm can be explained as follows:

1. Create the bounding box for each segment type. The bounding box is calculated for each segment and it is generated based on the minimum and maximum coordinates of the delta_x/y of the various routing costs computed in the previous step. Basically, each time a delta_x/y exceeds the bounding box for a specific segment type makes the bounding box to increase in size. This is done to have variable sizes for each segment type which may not require a cost map that spans over the whole size of the device if there is no need for it.
2. The cost entries from the routing expansions are stored within the cost_map divided in (delay and congestion)
3. Penalties are added for each segment type (see [comment above](https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1449#discussion_r482966378) that explains how the penalty works)
4. Filling the holes in the map (the routing costs may have holes when calculating the delays to the various delta_x, delta_y coordinates): this procedure happens in a way that the cost entry is filled using the nearby entries that are explored in a spiral manner so that the closest entries are being selected).

Reply:
Thanks @acomodi.  Point 1 is worth expanding:  (generated based on the minimum and maximum coordinates of the delay x/y of hte previous step).  I think I understand then that for some wire segments you don't create a map that is the size of the chip (dx,dy are smaller), but commenting why / how you choose to keep the bounding box smaller (and how you fill in the answer if you were ever asked for a bigger dx,dy for that segment in the lookahead) would be good.

Reply:
Commented all the steps in the below block of code